sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -266,141 +263,132 @@ public virtual async Task<Response<UploadLogsResult>> UploadAsync<T>(string rule
             {
-                // A list of tasks that are currently executing which will
-                // always be smaller than or equal to MaxWorkerCount
-                List<Task<BatchUpload>> runningTasks = new();
-                // Partition the stream into individual blocks
-                foreach (BatchedLogs<T> batch in Batch(logs, options))
+                try
                 {
                     // Start staging the next batch (but don't await the Task!)
-                    Task<BatchUpload> task = CommitBatchListSyncOrAsync(
+                    Task<Response> task = UploadBatchListSyncOrAsync(
                         batch,
                         ruleId,
                         streamName,
-                        true,
+                        async: true,
                         cancellationToken);
 
                     // Add the block to our task and commit lists
-                    runningTasks.Add(task);
+                    runningTasks.Add(Tuple.Create(task, batch.LogsList.Count));
 
                     // If we run out of workers
                     if (runningTasks.Count >= _maxWorkerCount)
                     {
                         // Wait for at least one of them to finish
-                        await Task.WhenAny(runningTasks).ConfigureAwait(false);
+                        await Task.WhenAny(runningTasks.Select(_ => _.Item1).ToList()).ConfigureAwait(false);
                         // Clear any completed blocks from the task list
                         for (int i = 0; i < runningTasks.Count; i++)
                         {
-                            Task<BatchUpload> runningTask = runningTasks[i];
+                            Task<Response> runningTask = runningTasks[i].Item1;
                             if (!runningTask.IsCompleted)
                             {
                                 continue;
                             }
-
-                            // Before removing runningTask - check if it has any errors and add to error list
-                            if (runningTask.Result.Error != null)
-                                errors.Add(runningTask.Result.Error);
-
+                            // If current task has an exception, log the exception and add number of logs in this task to failed logs
+                            if (runningTask.Exception != null)
+                            {
+                                exceptions ??= new List<Exception>();
+                                exceptions.Add(runningTask.Exception);
+                                logsFailed += runningTasks[i].Item2;
+                            }
+                            // If current task returned a response that was not a success, log the exception and add number of logs in this task to failed logs
+                            if (runningTask.Result.Status != 204)
+                            {
+                                exceptions ??= new List<Exception>();
```

Comment:
I think you can make this a single line:
```c#
exceptions ??= new List<Exception> { new Exception() };
```

Reply:
Nevermind.. that won't work for the case where the exceptions list is already instantiated.

But consider moving this into a separate method, e.g.:
```c#
private void AddToList(ref List<Exception> exceptions, Exception exception)
```